src/platform/cmx/boot_comp.c

```diff
@@ -0,0 +1,279 @@
+	 * region is also enabled. These are done by default */
+	chal_MPU_update(pgtbl, 1);
+	return 0;
+}
+
+//int
+//boot_pgtbl_mappings_add(struct captbl *ct, capid_t pgdcap, capid_t ptecap, const char *label,
+//		        void *kern_vaddr, unsigned long user_vaddr, unsigned int range, int uvm)
+//{
+//	u32_t count;
+//	int ret;
+//	u8_t *ptes;
+//	unsigned int nptes = 0, i;
+//	struct cap_pgtbl *pte_cap, *pgd_cap;
+//	pgtbl_t pgtbl;
+//
+//	pgd_cap = (struct cap_pgtbl*)captbl_lkup(ct, pgdcap);
+//	if (!pgd_cap || !CAP_TYPECHK(pgd_cap, CAP_PGTBL)) assert(0);
+//	pgtbl = (pgtbl_t)&(pgd_cap->pgtbl);
+////	nptes = boot_nptes(range);
+////	ptes  = mem_boot_alloc(nptes);
+////	assert(ptes);
+//
+//	printk("\tCreating %d %s PTEs for PGD @ 0x%x from [%x,%x) to [%x,%x).\n",
+//	       nptes, label, chal_pa2va((paddr_t)pgtbl),
+//	       kern_vaddr, kern_vaddr+range, user_vaddr, user_vaddr+range);
+//	/* Map in the two pages - we know that they start at 0x20020000, and end at 0x20030000 */
+//		pgtbl.data[0]=
+//	}
+//	/*
+//	 * Note the use of NULL here.  We aren't actually adding a PTE
+//	 * currently.  This is a hack and only used on boot-up.  We'll
+//	 * reuse this capability entry to create _multiple_ ptes.  We
+//	 * won't create captbl entries for each of them, so they
+//	 * cannot be aliased/removed later.  The only adverse
+//	 * side-effect I can think of from this is that we cannot
+//	 * reclaim all of the boot-time memory, but that is so far
+//	 * into the future, I don't think we care.
+//	 */
+//	/* PRY-TODO:add all the ptes here. in fact we only have four here */
+////	if (pgtbl_activate(ct, BOOT_CAPTBL_SELF_CT, ptecap, NULL, 1)) assert(0);
+////	pte_cap = (struct cap_pgtbl*)captbl_lkup(ct, ptecap);
+////	assert(pte_cap);
+////
+////	/* Hook in the PTEs */
+////	for (i = 0 ; i < nptes ; i++) {
+////		u8_t   *p  = ptes + i * PAGE_SIZE;
+////		paddr_t pf = chal_va2pa(p);
+////
+////		pgtbl_init_pte(p);
+////		//pte_cap->pgtbl = (pgtbl_t)p;
+////
+////		/* hook the pte into the boot component's page tables */
+////		ret = cap_cons(ct, pgdcap, ptecap, (capid_t)(user_vaddr + i*PGD_RANGE));
+////		assert(!ret);
+////	}
+////
+////	printk("\tMapping in %s.\n", label);
+////	/* Map in the actual memory. */
+////	for (i = 0 ; i < range/PAGE_SIZE ; i++) {
+////		u8_t *p     = kern_vaddr + i * PAGE_SIZE;
+////		paddr_t pf  = chal_va2pa(p);
+////		u32_t mapat = (u32_t)user_vaddr + i * PAGE_SIZE, flags = 0;
+////
+////		if (uvm  && pgtbl_mapping_add(pgtbl, mapat, pf, PGTBL_USER_DEF))
+////			assert(0);
+////		if (!uvm && pgtbl_cosframe_add(pgtbl, mapat, pf, PGTBL_COSFRAME))
+////			assert(0);
+////		assert((void*)p == pgtbl_lkup(pgtbl, user_vaddr+i*PAGE_SIZE, &flags));
+////	}
+//
+//	return 0;
+//}
+
+/* FIXME:  loops to create threads/tcaps/rcv caps per core. */
+static void
+kern_boot_thd(struct captbl *ct, void *thd_mem, void *tcap_mem)
```

Comment:
This one is largely generic and can be put into the kernel. Maybe we need a boot.c in the kernel.